### PR TITLE
Serialization and deserealization UGC structures to/from JSON

### PR DESCRIPTION
**Pull request by syershov**

### CODE REVIEW

File: 3party/jansson/myjansson.hpp [hunk 1]
```diff
@@ -36,9 +36,11 @@ class Json
 public:
   DECLARE_EXCEPTION(Exception, RootException);
 
-  explicit Json(std::string const & s) : Json(s.c_str()) {}
+  Json() = default;
+  explicit Json(std::string const & s) { Attach(s.c_str()); }
+  explicit Json(char const * s) { Attach(s); }
 
-  explicit Json(char const * s)
+  void Attach(char const * s)
```
Comment (kshalnev) [inline 3party/jansson/myjansson.hpp]: м.б. добавить еще и void Attach(const std::string& s) { Attach(s.c_str(); }?

File: ugc/serdes_json.hpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+namespace ugc
+{
+  template <typename Sink>
+  class SerializerJson
+  {
+  public:
+    SerializerJson(Sink & sink, HeaderV0 const & header) : m_sink(sink), m_header(header) {}
+
+    void operator()(bool const d, char const * name = nullptr)
```
Comment (kshalnev) [inline ugc/serdes_json.hpp]: можно enable_if использовать чтоб избавиться от copy-paste, например:
    template <typename ValueType, typename =
        typename std::enable_if<
            std::is_integral<ValueType>::value ||
            std::is_floating_point<ValueType>::value ||
            std::is_same<ValueType, std::string>::value
        >::type>
void operator()(ValueType const & d, char const * name = nullptr) {}
кстати, а float и double типы?

File: ugc/api.cpp [hunk 3]
```diff
@@ -11,7 +11,8 @@ namespace
 {
 Time FromDaysAgo(uint32_t days)
 {
-  return std::chrono::system_clock::now() - std::chrono::hours(days * 24);
+  return FromDaysSinceEpoch(
```
Comment (ygorshenin) [inline ugc/api.cpp]: Sorry, but what was wrong with old code?

File: ugc/serdes_json.hpp [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+  private:
+    json_t * SaveContext(char const * name = nullptr)
+    {
+      json_t * context = m_json;
+      if (name)
+        m_json = my::GetJSONObligatoryField(context, name);
+      return context;
+    }
+
+    void RestoreContext(json_t * context)
```
Comment (ygorshenin) [inline ugc/serdes_json.hpp]: As there is always a pair of SaveContext - RestoreContext, why not to:
```c++
template <typename Fn>
void GoToSubtree(Fn && fn)
{
  json_t * json = m_json;
  m_json = my::GetJSONObligatoryField(context, name);
  fn();
  m_json = json;  
}
```
and use it?

File: ugc/serdes_json.hpp [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+      for (auto const & v : vs)
+        (*this)(v);
+      Update(std::move(safe_json), name);
+    }
+
+    template <typename R>
+    void operator()(R const & r, char const * name = nullptr)
+    {
+      my::JSONPtr safe_json = std::move(m_json);
+      m_json = my::NewJSONObject();
```
Comment (ygorshenin) [inline ugc/serdes_json.hpp]: Same as below - is it possible to put all these safe_json saving and restoring into a helper function?

File: ugc/serdes_json.hpp [hunk 6]
```diff
@@ -0,0 +1,226 @@
+#pragma once
+
+#include "ugc/serdes.hpp"
```
Comment (ygorshenin) [inline ugc/serdes_json.hpp]: You don't use serdes here, right?

File: ugc/serdes_json.hpp [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+#include "ugc/types.hpp"
+
+#include "coding/multilang_utf8_string.hpp"
+#include "coding/point_to_integer.hpp"
```
Comment (ygorshenin) [inline ugc/serdes_json.hpp]: Where this is being used?

File: ugc/serdes_json.hpp [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+#include "coding/reader.hpp"
+#include "coding/varint.hpp"
+#include "coding/write_to_sink.hpp"
```
Comment (ygorshenin) [inline ugc/serdes_json.hpp]: Could you please point me why do you need this?

File: ugc/serdes_json.hpp [hunk 9]
```diff
@@ -0,0 +1,226 @@
+#pragma once
+
+#include "ugc/serdes.hpp"
+#include "ugc/types.hpp"
+
+#include "coding/multilang_utf8_string.hpp"
+#include "coding/point_to_integer.hpp"
+#include "coding/reader.hpp"
+#include "coding/varint.hpp"
```
Comment (ygorshenin) [inline ugc/serdes_json.hpp]: Don't see usage of varints here.

File: ugc/serdes_json.hpp [hunk 10]
```diff
@@ -0,0 +1,226 @@
+#pragma once
+
+#include "ugc/serdes.hpp"
+#include "ugc/types.hpp"
+
+#include "coding/multilang_utf8_string.hpp"
```
Comment (ygorshenin) [inline ugc/serdes_json.hpp]: Please, review your inclusion list.

File: ugc/serdes_json.hpp [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+#include "coding/point_to_integer.hpp"
+#include "coding/reader.hpp"
+#include "coding/varint.hpp"
+#include "coding/write_to_sink.hpp"
+
+#include "base/exception.hpp"
+
+#include "3party/jansson/myjansson.hpp"
+
+#include <cmath>
```
Comment (ygorshenin) [inline ugc/serdes_json.hpp]: Why do you need math here?

File: ugc/serdes_json.hpp [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+#include <cstdint>
+#include <iostream>
```
Comment (ygorshenin) [inline ugc/serdes_json.hpp]: Please, get rid of iostream.

File: ugc/ugc_tests/serdes_json_test.cpp [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+#include <cstdint>
+#include <vector>
+
+using namespace std;
+using namespace ugc;
+
+namespace
+{
+  using Buffer = vector<uint8_t>;
```
Comment (ygorshenin) [inline ugc/ugc_tests/serdes_json_test.cpp]: No need in 2-spaces indent.

File: ugc/ugc_tests/serdes_json_test.cpp [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+    HeaderV0 header;
+
+    Buffer buffer;
+
+    {
+      auto sink = MakeSink(buffer);
+      Ser ser(sink, header);
+      ser(expectedUGC);
+      ser.Flush();
```
Comment (ygorshenin) [inline ugc/ugc_tests/serdes_json_test.cpp]: See? I suggest you to move Flush() to dtor of serializer - in this case you may safely remove this line.

File: 3party/jansson/myjansson.hpp (hunk 1 above)
Comment (syershov) [inline 3party/jansson/myjansson.hpp]: done

File: ugc/serdes_json.hpp (hunk 2 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: I do it like in sedges.hpp, and won't change it.

File: ugc/api.cpp (hunk 3 above)
Comment (syershov) [inline ugc/api.cpp]: fixed

File: ugc/serdes_json.hpp (hunk 6 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/serdes_json.hpp (hunk 10 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/serdes_json.hpp (hunk 7 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/serdes_json.hpp (hunk 9 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/serdes_json.hpp (hunk 8 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/serdes_json.hpp (hunk 11 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/serdes_json.hpp (hunk 12 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/serdes_json.hpp [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+
+namespace ugc
+{
+  template <typename Sink>
+  class SerializerJson
```
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/serdes_json.hpp (hunk 2 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: We can't do this because of serialize vector of int, as example.

File: ugc/serdes_json.hpp (hunk 5 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/serdes_json.hpp [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+    void VisitRating(float const f, char const * name = nullptr)
+    {
+      CHECK_GREATER_OR_EQUAL(f, 0.0, ());
+      auto const d = static_cast<double>(f);
+      ToJSONObject(*m_json, name, d);
+    }
+
+    void Flush()
+    {
+      std::string json(json_dumps(m_json.get(), 0));
```
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/serdes_json.hpp [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+
+  template <typename Source>
+  class DeserializerJsonV0
+  {
+  public:
+    DECLARE_EXCEPTION(Exception, RootException);
+
+    DeserializerJsonV0(Source & source, HeaderV0 const & header) : m_source(source), m_header(header)
+    {
+      std::string src;
```
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/serdes_json.hpp [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    void Flush()
+    {
+      std::string json(json_dumps(m_json.get(), 0));
+      m_sink.Write(json.data(), json.size());
+    }
+
+  private:
+
```
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/ugc_tests/serdes_json_test.cpp (hunk 13 above)
Comment (syershov) [inline ugc/ugc_tests/serdes_json_test.cpp]: done

File: ugc/ugc_tests/serdes_json_test.cpp (hunk 14 above)
Comment (syershov) [inline ugc/ugc_tests/serdes_json_test.cpp]: done

File: ugc/serdes_json.hpp [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+    void VisitRating(float & f, char const * name = nullptr)
+    {
+      double d = 0.0;
+      FromJSONObject(m_json, name, d);
+      f = static_cast<float>(d);
+    }
+
+  private:
+    json_t * SaveContext(char const * name = nullptr)
```
Comment (syershov) [inline ugc/serdes_json.hpp]: Обсудили голосом, пока отказались.

File: ugc/serdes_json.hpp (hunk 4 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: Обсудили голосом, пока отказались.

**Comment by syershov** [comment]
@ygorshenin PTAL

File: ugc/ugc_tests/serdes_json_test.cpp (hunk 13 above)
Comment (ygorshenin) [inline ugc/ugc_tests/serdes_json_test.cpp]: Don't see where it's done.

File: ugc/serdes_json.hpp [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+
+  void VisitRating(float const f, char const * name = nullptr)
+  {
+    CHECK_GREATER_OR_EQUAL(f, 0.0, ());
+    auto const d = static_cast<double>(f);
+    ToJSONObject(*m_json, name, d);
+  }
+
+private:
+  void Update(my::JSONPtr json_object, char const * name, std::function<void()> && fn)
```
Comment (ygorshenin) [inline ugc/serdes_json.hpp]:
```c++
template <typename Fn>
void Update(..., Fn && fn)
```

Comment (ygorshenin) [inline ugc/serdes_json.hpp]: WithSubElem?

Comment (ygorshenin) [inline ugc/serdes_json.hpp]: WithNewScope?

Comment (ygorshenin) [inline ugc/serdes_json.hpp]: WithNewFrame?

Comment (syershov) [inline ugc/serdes_json.hpp]: done, NewScopeWith

**Comment by syershov** [comment]
@ygorshenin PTAL

File: ugc/serdes_json.hpp [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+namespace ugc
+{
+template <typename Sink>
+class SerializerJson
+{
+public:
+  SerializerJson(Sink & sink) : m_sink(sink) {}
+  ~SerializerJson()
+  {
+    std::string json(json_dumps(m_json.get(), 0));
```
Comment (ygorshenin) [inline ugc/serdes_json.hpp]: Memory leak here - the caller takes ownership over result from json_dumps.

File: ugc/ugc_tests/ugc_tests.pro [hunk 22]
```diff
@@ -19,3 +19,4 @@ QT *= core
 SOURCES += \
   ../../testing/testingmain.cpp \
   serdes_tests.cpp \
+
```
Comment (ygorshenin) [inline ugc/ugc_tests/ugc_tests.pro]: Git complains here about newline.

File: ugc/serdes_json.hpp (hunk 21 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/ugc_tests/ugc_tests.pro (hunk 22 above)
Comment (syershov) [inline ugc/ugc_tests/ugc_tests.pro]: done

**Comment by syershov** [comment]
@ygorshenin PTAL

File: ugc/serdes_json.hpp [hunk 23]
```diff
@@ -0,0 +1,180 @@
+#pragma once
+
+#include "ugc/types.hpp"
+
+#include "base/exception.hpp"
+
+#include "3party/jansson/myjansson.hpp"
+
+#include <cstdint>
+#include <stdlib.h>
```
Comment (ygorshenin) [inline ugc/serdes_json.hpp]: cstdlib, please.

File: ugc/ugc_tests/CMakeLists.txt [hunk 24]
```diff
@@ -9,6 +9,7 @@ omim_add_test(${PROJECT_NAME} ${SRC})
 omim_link_libraries(
   ${PROJECT_NAME}
   ugc
+  jansson
```
Comment (ygorshenin) [inline ugc/ugc_tests/CMakeLists.txt]: Please, move jansson after base but before stats_client.

File: ugc/serdes_json.hpp (hunk 23 above)
Comment (syershov) [inline ugc/serdes_json.hpp]: done

File: ugc/ugc_tests/CMakeLists.txt (hunk 24 above)
Comment (syershov) [inline ugc/ugc_tests/CMakeLists.txt]: done

**Comment by syershov** [comment]
@ygorshenin PTAL

**Pull request merged**